### ISSUE TITLE
Components only unsubscribe on update or unmount if they have custom update and unmount implementations.

### ISSUE THREAD

**Issue by CharlesStover** [CharlesStover/reactn#85]
The unsubscribe listener on update and unmount only binds to existing implementations. It does not fire if no implementation exists to which it can bind.

**Comment by umbertoghio** [comment]
Hello, 
I just run some tests on 2.0.3 but unfortunately the error persists, I tried to dig a bit deeper and I have more details:

Here the exact steps I take to reproduce:

- From Notifications page I touch an element, it will set notificationBadge in the global, then load the OffersDetail page
- OffersDetail page loads up just fine, page is mounted, also BottomBar is mounted
- BottomBar reads from this.global, so it subscribes to notificationBadge
- I go back to notifications, OffersDetail page is Unmounted, BottomBar is unmounted **but subscription remains? **
- I touch again the same notification, setGlobal fires, React calls a forceUpdate on BottomBar of the OfferDetails page, which is not mounted then the error

My theory is that due to stack navigation I have several Pages mounted at the same time, each page have his own BottomBar subscribed, so at any moment there are two / three BottomBar mounted, **but in this specific instance NOT the one in OfferDetails**

Questions that to my limited knowledge I can't answer, but maybe helpful:
How can reactn know which components to call a forceupdate ? 
Do the subscription considers the specific component with a unique key or id or just all the BottomBars ?

What is causing the unmount of offers details ?
I have a AppNavigation that has:
AppContainer -> switchNavigator -> DrawerNavigator -> StackNavigator -> (Notifications, offersDetails)
Switching from Notifications to OffersDetails will mount the offersDetails, going back to Notifications will unmount it

**Comment by umbertoghio** [comment]
I tried to setup a reproduction environment, I setup the navigation in the same way as my project but so far no luck...the main problem is that when going from Page A to Page B, then back to A, page B is kept mounted (in my project Page B unmounts)...can't figure out how to trigger a page B unmount.

I hope it can help....

https://github.com/umbertoghio/rn-test

**Comment by CharlesStover** [comment]
Thank you for the repo. I've cloned and ran it.

Does your `rn-test` repo cause the warning? I am unable to reproduce it, but I may not be looking in the right place. There are no warnings or errors displaying in my CLI or Expo Developer Tools screen.

Page B is mounting and unmounting fine and as expected. I've submitted a PR to your test repo with what I am running locally.

* Can you try removing your `node_modules` directory and re-installing them?
* If I am not looking in the right place to see the error message, can you direct me to where to find it?

**Comment by umbertoghio** [comment]
Ho Charles,
Sorry, I could have described the issue 
better.
The repo provided is as close as possible to my work repo, but I can’t reproduce the 
problem.

There’s one key difference I can’t explain: 
In this test repo pages remain mounted after navigation, while on my work repo page is unmounted (therefore the error).
I’ll try to investigate more

**Comment by CharlesStover** [comment]
For me, the pages are unmounting correctly on navigation.

Your `console.log('PageName is unmounting')` was inside the `componentWillMount` method, as opposed to the `componentWillUnmount` method. This may be why you believed navigating did not result in an unmount. You were not logging anything on unmount.

I gave those methods the correct names in my PR.

Your work repo is on `reactn@2.0.3` and still has the "cannot re-render an unmounted Component" warning?

**Comment by umbertoghio** [comment]
Wops too much copy pasting....
yes, no errors, I’ll fix it and give my feedback

**Comment by umbertoghio** [comment]
Ok, tested your changes, thank you for the correction, I confirm it formally behaves like my main project (mount / unmount), but I get no errors or warnings, even on version 2.0.0.

Just to test I tried to add a componentwillunmount to my main project bar but still get the error.

I'll try a different approach: I'll start to prune and chop down my main project until I figure out the culprit, but I need a bit of time.

Thank you very much with all the efforts and assistance!

**Comment by CharlesStover** [comment]
No, no, thank you very much for assisting in the reproduction.

Let me know when your repository reproduces the error, and I'll start working on a fix immediately. :)

**Comment by umbertoghio** [comment]
Eureka, I made it :-)

I started from the main project and started to chop down everything that was not essential in reproducing the problem, ended up with a very simple reproduction scenario that triggers the problem!

Note that I have the issue on many pages / components but I hope this use case will allow you to track down the issue :-)

Repo is still the same
https://github.com/umbertoghio/rn-test

Thanks a lot!

PS : in case I want to offer you a beer where I can PayPal?

**Comment by umbertoghio** [comment]
Hi Charles, 
Just an update, we have a companion React (last version) web project to the app and noticed that when upgrading to reactn 2.x the error is triggered also there, so it is not just something from react native.
Do you need a reproduction scenario for the web too?
Can you reopen this issue in the mean time?
Thanks!


<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by CharlesStover** [comment]
I'll try testing on the React Native project this weekend. A reproduction in web might help. At the very least, it will be faster and easier to test than Native.

**Comment by CharlesStover** [comment]
I'm not particularly familiar with React Native or your navigation library, but I'll comment this context I've discovered as far as the rendering behavior of your repository:

The first two things that occur:
BarBottom mounted.
Notifications mounted.

On navigation change:
BarBottom mounted.
OfferDetails mounted.
Important note: The previous two Components did not unmount here.

On navigation change:
OfferDetails unmounting.
BarBottom unmounting.
Important note: The first two Components did not re-mount, because they were already mounted.

On navigation change:
BarBottom mounted.
OfferDetails mounted.
Warning occurs.

The fact that the first two Components never unmount or re-mount is unexpected behavior. It is hard to debug whether or not the warning is a result of the ReactN package or the improper rendering lifecycle of those Components, which may be due to the navigation package or other factors.

I'll keep investigating, but I wanted to document the above as a part of this debugging process.

**Comment by umbertoghio** [comment]
Hi Chales, 
thank you for your insights, for what I understand the StackNavigator keeps the "pages" mounted to allow to press the back button and go back to the previous page, so it seems that navigation is not at fault: https://reactnavigation.org/docs/en/navigation-lifecycle.html

What I think is strange is that in Reactn 1.0.0 there are no errors, while they appear from version 2.0.0..is something important changed in the way subscription to changes is handled?

I'm setting up a repro environment in react web, I have the same warning when running this code triggering a page change:

![Screen Shot 2019-06-09 at 11 26 16](https://user-images.githubusercontent.com/25666241/59157347-7e06ed00-8aa9-11e9-889d-edec6e65f0da.png)

thank you very much

**Comment by umbertoghio** [comment]
Some progress, while slimming down my project to arrive at a minimum code scenario I noticed something in common with the other react-native project:

The problem seems to occur when reading from TWO variables from global, i.e.:

![Screen Shot 2019-06-09 at 12 27 46](https://user-images.githubusercontent.com/25666241/59157836-01c4d780-8ab2-11e9-9170-b4cb5454097f.png)

this won't trigger the issue...

![Screen Shot 2019-06-09 at 12 28 40](https://user-images.githubusercontent.com/25666241/59157843-215c0000-8ab2-11e9-8ef4-c098d1ce973e.png)

**Comment by umbertoghio** [comment]
Ok, I'm getting good at this :-)

Reproduction environment in just React Web here: 

https://github.com/umbertoghio/reactn-withjustreact/tree/master/src

the logic is very similar to the one in react-native, issue is somehow involving 

- react-navigation
- reading TWO items from the global state
- setting those two items from the other page componentdidmount

Hope that helps!

Umberto

**Comment by CharlesStover** [comment]
Thank you for a web reproduction. I am definitely able to make faster progress with it.

This is actually a different issue than the OP, but I will resolve it here for the sake of documentation of the above steps and repositories.

You were very accurate that this problem occurs when subscribing to two properties. The underlying root cause appears to be that, when unsubscribing from the global state on unmount, only the first property is unsubscribed. If a Component is subscribed to more than one property, it will remain subscribed to at least one property.

I've added the following unit test, which verified this error (currently fails), and will deploy a fix in `2.1.3` today:

```javascript
  it('should remove more than 1 property listener', (): void => {

    // Create a global state manager that has one property listener (one Component subscription) on two properties.
    globalStateManager.addPropertyListener(PROPERTY, PROPERTY_LISTENER);
    globalStateManager.addPropertyListener(PROPERTY2, PROPERTY_LISTENER);
    expect(globalStateManager.hasPropertyListener(PROPERTY_LISTENER))
      .toBe(true);

    // Remove that property listener from the global state.
    globalStateManager.removePropertyListener(PROPERTY_LISTENER);

    // Expect that the property listener no longer exists on any properties in the global state.
    expect(globalStateManager.hasPropertyListener(PROPERTY_LISTENER))
      .toBe(false);
  });
```

Currently, the property listener still exists on the second property and only is removed from the first.

Thanks a ton for your help! It was invaluable. 😁

**Comment by umbertoghio** [comment]
Hi Charles, 
thanks a million, tested the React Native app and it's working flawlessly, I can finally use ReactN-devtools :-)
